uips/uip-4.md

```diff
@@ -0,0 +1,102 @@
+# UIP: Spend Backreferences
+
+| uip | 4 |
+| - | - |
+| title | Spend Backreferences |
+| description | Spend Backreferences enable improved sync performance |
+| author | Jennifer Helsby ([@redshiftzero](https://github.com/redshiftzero)), Henry de Valence ([@hdevalence](https://github.com/hdevalence)), Lúcás Meier ([@cronokirby](https://github.com/cronokirby))|
+| discussions-to | <https://forum.penumbra.zone/t/uip-spend-backreferences/110> |
+| status | Draft |
+| type | Standards Track |
+| consensus | Yes |
+| created | 2024-11-06 |
+
+## Abstract
+
+This specification introduces a method to improve Penumbra sync speeds by adding additional data that can be used by DAGSync clients. `Spend` actions will contain a new `backref_commitment` field, allowing clients to traverse their transaction graph backwards and quickly recover their entire transaction history.
+
+## Motivation
+
+DAGSync is a graph-aware fast syncing algorithm. A client, upon detecting a single transaction involving them, can check that outputs visible to them are spent or not. If the output is unspent, then they have identified a live note they can potentially spend in the future, else if the output is unspent, they can continue the process forwards in the transaction graph, until they reach unspent notes.
+
+The design of Penumbra currently does not allow traversal backwards through the transaction graph, only forwards. A `Spend` intentionally does not reveal the note being spent, only the nullifier that is revealed. By including on the `Spend` an encrypted reference back to the note commitment being spent, such that only the note owner can view it, we enable DAGSync clients to efficienctly reconstruct the transaction history both backwards and forwards.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Modification to `SpendBody`
+
+The `Spend` action will be augmented with an additional field `backref_commitment` on the `SpendBody`:
```

Comment:
I think my original naming suggestion was a bad one, because there are really two different things here: the commitment to the note the action spends, and the encryption of that commitment. If we call the encryption `backref_commitment` then we either (1) invent a new name for that commitment [confusing] or (2) have no distinction in naming [also confusing].

Instead I think it would be better to call it `encrypted_backref`, so that we aren't confusing the plaintext and ciphertext.